crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,126 @@
+use std::{marker::PhantomData, ops::DerefMut};
+
+use bevy_app::Plugin;
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    bundle::Bundle,
+    component::Component,
+    entity::{Entity, EntityHashMap},
+    observer::Trigger,
+    query::With,
+    reflect::ReflectComponent,
+    system::{Query, ResMut, Resource},
+    world::{Mut, OnAdd, OnRemove, World},
+};
+use bevy_hierarchy::DespawnRecursiveExt;
+use bevy_reflect::Reflect;
+
+/// Marker component that indicate its entity needs to be Synchronized to render world
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[reflect[Component]]
+pub struct ToRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug)]
+#[component(storage = "SparseSet")]
+pub struct RenderEntity(Entity);
+impl RenderEntity {
+    pub fn entity(&self) -> Entity {
+        self.0
+    }
+}
+
+#[derive(Component, Deref, Clone, Debug)]
+#[component(storage = "SparseSet")]
+pub struct MainEntity(Entity);
+impl MainEntity {
+    pub fn entity(&self) -> Entity {
+        self.0
+    }
+}
+// marker component that its entity needs to be despawned per frame.
+#[derive(Component, Clone, Debug, Default, Reflect)]
+pub struct RenderFlyEntity;
+
+pub(crate) enum EntityRecord {
+    Added(Entity),
+    // (main , render)
+    Removed(Entity, Entity),
+}
+
+// Entity Record in MainWorld pending to Sync
+#[derive(Resource, Default, Deref, DerefMut)]
+pub(crate) struct PendingSyncEntity {
+    records: Vec<EntityRecord>,
+}
+
+// resource to maintain entity mapping from the main world to the render world
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct MainToRenderEntityMap {
+    map: EntityHashMap<Entity>,
+}
+
+pub(crate) fn entity_sync_system(main_world: &mut World, render_world: &mut World) {
+    render_world.resource_scope(|render_world, mut mapper: Mut<MainToRenderEntityMap>| {
+        let mapper = mapper.deref_mut();
+        main_world.resource_scope(|world, mut pending: Mut<PendingSyncEntity>| {
+            for record in pending.drain(..) {
+                match record {
+                    EntityRecord::Added(e) => {
+                        if let Some(mut entity) = world.get_entity_mut(e) {
+                            match entity.entry::<RenderEntity>() {
+                                bevy_ecs::world::Entry::Occupied(_) => {}
+                                bevy_ecs::world::Entry::Vacant(entry) => {
+                                    let id = render_world.spawn(MainEntity(e)).id();
+
+                                    mapper.insert(e, id);
+                                    entry.insert(RenderEntity(id));
+                                }
+                            };
+                        }
+                    }
+                    EntityRecord::Removed(e1, e2) => {
+                        mapper.remove(&e1);
```

Comment:
Won't this cause entities to be deleted prematurely if a single one of their components is removed? If I understand correctly OnRemove triggers when any of the components in its bundle is removed, right? If this is a problem I'm assuming it hasn't been caught because out internal systems often despawn the whole entity at once.

Reply:
We might want to have something more fine-grained here.

Reply:
if you haven't added tracking for a component (using WorldSyncPlugin<Component) , the removal of this component will not trigger this event.

Reply:
That's not the issue, unless I'm misunderstanding. Currently even if you remove a single component it'll delete the render world entity it seems like?

Reply:
The synchronization relationship will  lost only when you delete a specific component . This loss of synchronization is confined to the lifetime scope of that particular component.

For example
e1   [Camera,Camera3D,VisibleEntities,*] 

Because we only added WorldSyncPlugin<Camera>, the synchronization relationship will persist even if Camera3D or VisibleEntities are removed

Reply:
What's the behavior if two synchronized components are on the same entity, and one is removed? It may not be the "canon" use of `WorldSyncPlugin` but it probably shouldn't be *disallowed*.